mlrun/runtimes/pod.py

```diff
@@ -411,6 +420,264 @@ def enrich_resources_with_default_pod_resources(
+            return
+
+        # Only add non-matching tolerations to avoid duplications
+        for function_toleration in self.tolerations:
+            for toleration in tolerations:
+                if function_toleration != toleration:
+                    tolerations_to_add.append(toleration)
+
+        if len(tolerations_to_add) > 0:
+            self.tolerations += tolerations_to_add
+
+    def enrich_function_preemption_spec(self):
+        """
+        Enriches function pod with the below described spec. if no platformConfiguration related
+        configuration is given, do nothing.
+            `Allow` 	- Adds Tolerations if taints were given.
+                        otherwise, assume pods can be scheduled on preemptible nodes.
+                        > Purges any `affinity` / `anti-affinity` preemption related configuration
+            `Constrain` - Uses node-affinity to make sure pods are assigned using OR on the given node label selectors.
+                        > Uses `Allow` configuration as well.
+                        > Purges any `anti-affinity` preemption related configuration
+            `Prevent`	- Prevention is done either using taints (if Tolerations were given) or anti-affinity.
+                        > Purges any `tolerations` / `gpuTolerations` preemption related configuration
+                        > Purges any `affinity` preemption related configuration
+                        > Adds anti-affinity IF no tolerations were given
+        """
+        # nothing to do here, configuration is not populated
+        if (
+            not mlconf.get_preemptible_tolerations()
+            and not mlconf.get_preemptible_node_selector()
+        ):
+            return
+
+        if not self.preemption_mode:
+            self.preemption_mode = mlconf.default_preemption_mode
+            logger.debug(
+                "No preemption mode was given, using the default preemption mode",
+                new_preemption_mode=self.preemption_mode,
+            )
+
+        logger.debug(
+            "Enriching function spec for given preemption mode",
+            preemption_mode=self.preemption_mode,
+        )
+        # remove preemptible tolerations and remove preemption related configuration
+        # and enrich with anti-affinity if preemptible tolerations configuration haven't been provided
+        if self.preemption_mode == PreemptionModes.prevent:
+            # ensure no preemptible node tolerations
+            self._prune_tolerations(mlconf.get_preemptible_tolerations())
+            # if preemptible tolerations were given, purge affinity preemption related configuration
+            if mlconf.get_preemptible_tolerations():
+                self._prune_affinity_node_selector_requirement(
+                    (
+                        compile_affinity_by_label_selector(
+                            NodeSelectorOperator.node_selector_op_in
+                        )
+                    )
+                )
```

Comment:
The main reason is that if there are preemptible tolerations and we prune them, pod can't be scheduled on the preemptible nodes because it doesn't tolerates the preemptible nodes. On the other side when preemptible nodes aren't tainted then we need to add anti-affinity.